src/lib/Builder.hs

```diff
@@ -844,27 +839,39 @@ zeroAt ty = case ty of
 zeroLike :: (HasCallStack, HasType e, Builder m) => e n -> m n (Atom n )
 zeroLike x = zeroAt =<< getType x
 
-tangentType :: Type n -> Type n
-tangentType ty = case maybeTangentType ty of
-  Just tanTy -> tanTy
+tangentType :: EnvReader m => Type n -> m n (Type n)
+tangentType ty = maybeTangentType ty >>= \case
+  Just tanTy -> return tanTy
   Nothing -> error $ "can't differentiate wrt type: " ++ pprint ty
 
-maybeTangentType :: Type n -> Maybe (Type n)
-maybeTangentType ty = case ty of
-  StaticRecordTy items -> StaticRecordTy <$> mapM maybeTangentType items
-  TypeCon _ _ _ -> Nothing -- Need to synthesize or look up a tangent ADT
+maybeTangentType :: EnvReader m => Type n -> m n (Maybe (Type n))
+maybeTangentType ty = liftEnvReaderT $ maybeTangentType' ty
+
+maybeTangentType' :: Type n -> EnvReaderT Maybe n (Type n)
+maybeTangentType' ty = case ty of
+  StaticRecordTy items -> StaticRecordTy <$> mapM rec items
+  TypeCon _ _ _ -> rec =<< fromNewtypeWrapper ty
```

Comment:
Hmmm this feels quite ad-hoc. Do we really need to be able to differentiate wrt index ranges?

Reply:
Yes, it's ad-hoc, but I don't think I'm ready to tackle arbitrary user-defined ADTs yet. We don't differentiate wrt index ranges directly, but the tests do include cases where we differentiate functions that create intermediate index range variables, and we to be able to evaluate their (trivial) tangent type in order to decide that they're trivial wrt AD.